src/mpi/romio/adio/ad_quobytefs/ad_quobytefs.c

```diff
@@ -0,0 +1,116 @@
+    ADIOI_FAKE_IwriteContig,    /* IwriteContig */
+#endif
+    ADIOI_GEN_IODone,   /* ReadDone */
+    ADIOI_GEN_IODone,   /* WriteDone */
+    ADIOI_GEN_IOComplete,       /* ReadComplete */
+    ADIOI_GEN_IOComplete,       /* WriteComplete */
+    ADIOI_GEN_IreadStrided,     /* IreadStrided */
+    ADIOI_GEN_IwriteStrided,    /* IwriteStrided */
+    ADIOI_QUOBYTEFS_Flush,      /* Flush */
+    ADIOI_QUOBYTEFS_Resize,     /* Resize */
+    ADIOI_QUOBYTEFS_Delete,     /* Delete */
+    ADIOI_GEN_Feature,  /* Features */
+    "QUOBYTEFS:ROMIO driver for quobyte file system",
+    ADIOI_GEN_IreadStridedColl, /* IreadStridedColl */
+    ADIOI_GEN_IwriteStridedColl,        /* IwriteStridedColl */
+    ADIOI_QUOBYTEFS_SetLock     /* SetLock */
+};
+
+char *ADIOI_QUOBYTEFS_ExtractRegistry(const char *filename, int *error_code)
+{
+    // input: //registry.address/[volume/]path
+    // output: registry.address
+    static char myname[] = "ADIOI_QUOBYTEFS_ExtractRegistry";
+    const char *prefix = "//";
+    int prefix_size = strlen(prefix);
+    if (!strncmp(filename, prefix, prefix_size)) {
+        *error_code = MPIO_Err_create_code(MPI_SUCCESS, MPIR_ERR_FATAL, myname,
+                                           __LINE__, MPI_ERR_NAME, "Invalid uri", 0);
+    }
+    char *extract_filename = (char *) filename + prefix_size;
+    char *tmp = strchr(extract_filename, '/');
+    char *registry = NULL;
+
+    if (tmp != NULL && tmp > extract_filename) {
+        size_t length = tmp - extract_filename;
+        registry = strndup(extract_filename, length);
+        *error_code = MPI_SUCCESS;
+        return registry;
+    } else {
+        *error_code = ADIOI_Err_create_code(myname, filename, EINVAL);
+    }
+    return NULL;
+}
+
+void ADIOI_QUOBYTEFS_CreateAdapter(const char *filename, int *error_code)
+{
+    static char myname[] = "ADIOI_QUOBYTEFS_CreateAdapter";
+
+    char *registry = ADIOI_QUOBYTEFS_ExtractRegistry(filename, error_code);
+    if (registry == NULL || *error_code != MPI_SUCCESS) {
+        return;
+    }
+    const char process_name[] = "adio_ffffffff";
+    char name_buffer[strlen(process_name)];
+    int rank;
+    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
+    snprintf(name_buffer, strlen(process_name), "adio_%x", rank);
+    quobyte_set_process_name(name_buffer);
+    int create_status = quobyte_create_adapter(registry);
+    free(registry);
+    if (create_status != EISCONN && create_status != 0) {
+        *error_code = MPIO_Err_create_code(MPI_SUCCESS, MPIR_ERR_FATAL, myname,
+                                           __LINE__, MPI_ERR_IO,
+                                           "Could not create quobyte adapter", 0);
+        return;
+    }
+    global_quobyte_io_context = -1;
+}
+
+void ADIOI_QUOBYTEFS_DestroyAdapter()
+{
+    //TODO(alexey): place holder adapter destruction,
+    //              for when it works as expected
+}
+
+const char *GetVolumeAndPath(const char *filename)
```

Comment:
Should this be a static function?

Reply:
It is used in ad_quobytefs_delete.c, ad_quobytefs_fcntl.c and ad_quobytefs_open.c
Do you want me to implement it static in the header instead?

Reply:
There are two possibilities.  It can be a static in a header (but with an unused attribute so files that include the header but don't use it don't complain), or it could be moved to an internal header file and the function namespaced accordingly.  The `GetVolumeAndPath` symbol shouldn't be exposed.  It'll need to come under the `ADIOI_` namespace somehow.  Perhaps as `ADIOI_QUOBYTEFSI_`?